src/tools/w3dviewer/emitterpropertysheet.cpp

```diff
@@ -0,0 +1,198 @@
+        document->Deselect();
+        m_name = m_instanceList->Get_Name();
+        m_instanceList->Clear_Emitters();
+        document->AddEmittersToList(m_instanceList, m_instanceList->Get_Name(), nullptr);
+    }
+}
+
+void EmitterPropertySheetClass::Initialize()
+{
+    if (m_instanceList != nullptr) {
+        m_name = m_instanceList->Get_Name();
+    } else {
+        CreateEmitter();
+    }
+
+    m_generalPage.SetInstanceList(m_instanceList);
+    m_generalPage.Initialize();
+    m_particlePage.SetInstanceList(m_instanceList);
+    m_particlePage.Initialize();
+    m_physicsPage.SetInstanceList(m_instanceList);
+    m_physicsPage.Initialize();
+    m_colorPage.SetInstanceList(m_instanceList);
+    m_colorPage.Initialize();
+    m_userPage.SetInstanceList(m_instanceList);
+    m_userPage.Initialize();
+    m_sizePage.SetInstanceList(m_instanceList);
+    m_sizePage.Initialize();
+    m_linePage.SetInstanceList(m_instanceList);
+    m_linePage.Initialize();
+    m_rotationPage.SetInstanceList(m_instanceList);
+    m_rotationPage.Initialize();
+    m_framePage.SetInstanceList(m_instanceList);
+    m_framePage.Initialize();
+    m_lineGroupPage.SetInstanceList(m_instanceList);
+    m_lineGroupPage.Initialize();
+
+    AddPage(&m_generalPage);
+    AddPage(&m_particlePage);
+    AddPage(&m_physicsPage);
+    AddPage(&m_colorPage);
+    AddPage(&m_sizePage);
+    AddPage(&m_userPage);
+    AddPage(&m_linePage);
+    AddPage(&m_rotationPage);
+    AddPage(&m_framePage);
+    AddPage(&m_lineGroupPage);
+    m_generalPage.SetPropertySheet(this);
+}
+
+void EmitterPropertySheetClass::CreateEmitter()
+{
+    ParticlePropertyStruct<Vector3> color;
+    memset(&color, 0, sizeof(color));
+    color.Start = Vector3(1.0f, 1.0f, 1.0f);
+    ParticlePropertyStruct<float> opacity;
+    memset(&opacity, 0, sizeof(opacity));
+    opacity.Start = 1.0f;
+    ParticlePropertyStruct<float> size;
+    memset(&size, 0, sizeof(size));
+    size.Start = 0.1f;
+    ParticlePropertyStruct<float> rotation;
+    memset(&rotation, 0, sizeof(rotation));
+    ParticlePropertyStruct<float> frames;
+    memset(&frames, 0, sizeof(frames));
+    ParticlePropertyStruct<float> blurtimes;
+    memset(&blurtimes, 0, sizeof(blurtimes));
+
+    ParticleEmitterClass *emitter = new ParticleEmitterClass(10.0f,
+        1,
+        new Vector3SolidBoxRandomizer(Vector3(0.1f, 0.1f, 0.1f)),
+        Vector3(0.0f, 0.0f, 1.0f),
+        new Vector3SolidBoxRandomizer(Vector3(0.0f, 0.0f, 1.0f)),
+        0.0f,
+        0.0f,
+        color,
+        opacity,
+        size,
+        rotation,
+        0.0f,
+        frames,
+        blurtimes,
+        Vector3(0.0f, 0.0f, 0.0f),
+        1.0f,
+        0.0f,
+        nullptr);
+
+    m_instanceList = new EmitterInstanceList();
+    m_instanceList->Add_Emitter(emitter);
+    GetCurrentDocument()->SetParticleEmitter(emitter, true, true);
+    Ref_Ptr_Release(emitter);

```

Comment:
This looks suspicious, because there was no Add_Ref. I suspect new EmitterInstanceList starts with ref count 1 then and Add_Emitter adds 1 more? Otherwise it makes no sense.

Reply:
Yes this is correct.